### PR TITLE
Update `django.template.base.Template.render()` argument type

### LINKED ISSUE #1147: django.template.Template.render() has incorrectly typed argument
# Bug report

## What's wrong

`django.template.Template.render()` has the following stub (see [here](https://github.com/typeddjango/django-stubs/blob/master/django-stubs/template/base.pyi#L76)):

```python
def render(self, context: Optional[Union[Context, Dict[str, Any]]]) -> SafeString: ...
```

However, the implementation of this method below makes it impossible to pass in a `Dict[str, Any]` (see [here](https://github.com/django/django/blob/main/django/template/base.py#L169)) as it immediately attempts to access an attribute `render_context`:

```python
def render(self, context):
    "Display stage -- can be called many times"
    with context.render_context.push_state(self):
        if context.template is None:
            with context.bind_template(self):
                context.template_name = self.name
                return self._render(context)
        else:
            return self._render(context)
```

From the looks of it, the various backends define their own `Template` class, some (all?) of which have a `render()` method that accepts dictionaries, but this is not the case for the standard `django.template.Template` class.

I don't know enough of the surrounding context to say whether this is intended or not, but it seems wrong from a first glance. Apologies if this has already been answered, I did have a look through the issues on the repo to try to find similar queries.

## How is that should be

The stub for `django.template.Template.render()` should be changed to:

```python
def render(self, context: Optional[Context]) -> SafeString: ...
```

## System information

- OS: macOS 12.2.1
- `python` version: 3.10.6
- `django` version: 4.0.7
- `mypy` version: 0.971
- `django-stubs` version: 1.12.0
- `django-stubs-ext` version: 0.5.0

### PR DESCRIPTION
**Pull request by Majsvaffla**
# I have made things!

I have corrected the type definition for the _context_ argument to `django.template.base.Template.render`. The argument is no longer typed as optional and does no longer accept a _dict_ value.

## Related issues

- Closes #1147

### CODE REVIEW

**Review by sobolevn — APPROVED** [review]
I am quite unsure what to do.

Removing this completely means that older versions of django become unsupported. 

@adamchainz what do you think?

**Comment by sobolevn** [comment]
(I meant to click "Comment", not "Approve")

**Comment by adamchainz** [comment]
I'd like to make some hints conditional by Django version, if that could be possible... perhaps the mypy plugin could provide the current version?

**Comment by sobolevn** [comment]
As far as I know - custom builtins are not supported. https://github.com/python/mypy/issues/12172

We can do something like:

```python
_DjangoVersion4_1 = _DjangoVersion[4, 1]
```

Inside our plugin we can return `Literal[True]` for this using `get_type_analyze_hook()` from https://mypy.readthedocs.io/en/stable/extending_mypy.html

But, this is quite complex!

**Comment by adamchainz** [comment]
I was thinking about typing `django.VERSION` conditionally, so we could use the same syntax as at runtime:

```py
if django.VERSION >= (4 ,1):
    ...  # new stuff
else:
    ...  # old stuff
```

It seems though that Mypy cannot type narrow on comparisons of literal ints in tuples, like below:

```py
from typing import Literal
from typing_extensions import assert_never

version: tuple[Literal[4], Literal[1]] = (4, 1)

if version >= (4, 1):
    reveal_type(version)
else:
    assert_never(version)
```

It doesn't work:

```
$ mypy example.py
example.py:7: note: Revealed type is "Tuple[Literal[4], Literal[1]]"
example.py:9: note: Revealed type is "Tuple[Literal[4], Literal[1]]"
example.py:10: error: Argument 1 to "assert_never" has incompatible type "Tuple[Literal[4], Literal[1]]"; expected "NoReturn"
Found 1 error in 1 file (checked 1 source file)
```

It seems this limitation also applies to literal ints in general:

```py
from typing import Literal
from typing_extensions import assert_never

version: Literal[4] = 4

if version >= 4:
    reveal_type(version)
else:
    reveal_type(version)
    assert_never(version)
```

Perhaps the better approach is as you suggested, some alternative true/false constants. Perhaps we fix to the latest version but override that in the plugin based on the installed version of django that we find, or some config.

**Comment by sobolevn** [comment]
`sys.version_info` check is hard-coded. This is why I thought about `Literal[True]` / `Literal[False]` hacks with analyzed dynamic types.

**Comment by ngnpope** [comment]
Also found this related issue: https://github.com/python/typing/issues/693

Following on from Adam's example, I did note that `==` and `!=` with a literal number do work as expected:
```python
from typing import Literal
from typing_extensions import assert_never

version: Literal[4] = 4

if version == 4:
    reveal_type(version)
else:
    reveal_type(version)
    assert_never(version)

if version != 4:
    reveal_type(version)
    assert_never(version)
else:
    reveal_type(version)

if version == 4 or version == 5:
    reveal_type(version)
else:
    reveal_type(version)
    assert_never(version)
```
This yields the following:
```console
❯ mypy test.py 
test.py:7: note: Revealed type is "Literal[4]"
test.py:16: note: Revealed type is "Literal[4]"
test.py:19: note: Revealed type is "Literal[4]"
Success: no issues found in 1 source file
```
So you could possibly construct something very verbose to achieve this by exposing major and minor versions as independent variables and thrashing out all combinations.

I'm currently looking into whether we can extend `mypy` to allow for checking additional things as well as `sys.version_info` when inferring condition values. It looks like it should be relatively straightforward to piggy back off the existing logic and provide a callback... I'll have to have a go.

**Comment by adamchainz** [comment]
> So you could possibly construct something very verbose to achieve this by exposing major and minor versions as independent variables and thrashing out all combinations.

We could also use a single numbers: `41` for Django 4.1, `42` for Django 4.2, etc.

> I'm currently looking into whether we can extend `mypy` to allow for checking additional things as well as `sys.version_info` when inferring condition values. It looks like it should be relatively straightforward to piggy back off the existing logic and provide a callback... I'll have to have a go.

Sick

File: django-stubs/template/base.pyi [hunk 1]
```diff
@@ -73,7 +73,7 @@ class Template:
         engine: Optional[Engine] = ...,
     ) -> None: ...
     def __iter__(self) -> Iterator[Node]: ...
-    def render(self, context: Optional[Union[Context, Dict[str, Any]]]) -> SafeString: ...
+    def render(self, context: Context) -> SafeString: ...
```
Comment (intgr) [inline django-stubs/template/base.pyi:60]: Note that this has to be compatible with `django.template.backends.base._EngineTemplate` protocol defined in django-stubs.

Comment (intgr) [reply django-stubs/template/base.pyi:60]: Oh, wait. There are two `Template` classes in Django

`django.template.base.Template.render()` doesn't support dict, but `django.template.backends.django.Template.render()` is a wrapper class that does support `dict` and `None`.

`django.template.backends.django.DjangoTemplates.get_template()` wraps the first Template in the latter wrapper class. Confusing!

**Comment by intgr** [comment]
> Removing this completely means that older versions of django become unsupported.

Note that this is not as bad as it actually looks.

If you get your template via the `django.template.loader.get_template()` or `get_template()` method of the template engine, you still get a Template whose `render()` supports `dict` and `None`.

As noted in my other comment:

> Oh, wait. There are two `Template` classes in Django
> 
> `django.template.base.Template.render()` doesn't support `dict`, but `django.template.backends.django.Template.render()` is a wrapper class that does support `dict` and `None`.
> 
> `django.template.backends.django.DjangoTemplates.get_template()` wraps the first Template in the latter wrapper class. Confusing!

**Comment by christianbundy** [comment]
> We could also use a single numbers: 41 for Django 4.1, 42 for Django 4.2, etc.

Yep! Or a NamedTuple if we really needed this specificity:

```python3
from typing import Literal, NamedTuple
from typing_extensions import assert_never


class Version(NamedTuple):
    major: Literal[4]
    minor: Literal[1]
    
version = Version(4, 1)

if version.major == 4:
    reveal_type(version.major)
else:
    reveal_type(version.major)
    assert_never(version)

if version.major != 4:
    reveal_type(version.major)
    assert_never(version)
else:
    reveal_type(version.major)

if version.major == 4 or version.major == 5:
    reveal_type(version.major)
else:
    reveal_type(version.major)
    assert_never(version)
```

**Review by adamchainz — APPROVED** [review]
I think we can merge this one now, as in https://github.com/typeddjango/django-stubs/issues/1095#issuecomment-1307066177 we agreed to support just the latest Django version.

**Comment by adamchainz** [comment]
I opened an issue about conditional types based on Django version: #1244

**Comment by sobolevn** [comment]
Let's merge it right after `1.13.1` release

**Comment by Majsvaffla** [comment]
I haven't followed along completely but if I understand correctly, I could resolve the conflicts and this is still to be merged?

**Comment by sobolevn** [comment]
Yes :)

**Comment by flaeppe** [comment]
Thanks for the PR and sorry for being so super slow getting it merged

**Pull request merged**